executor heartbeat meets [DFLOW:ErrUnknownExecutorID] and exits

ref: https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true

```go
server-executor-0_1      | 2022-04-28T03:49:58.891976512Z [2022/04/28 03:49:58.890 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-0:10240]
server-executor-0_1      | 2022-04-28T03:49:58.901128619Z [2022/04/28 03:49:58.900 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-1:10240]
server-executor-0_1      | 2022-04-28T03:49:58.905541022Z [2022/04/28 03:49:58.904 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-2:10240]
server-executor-0_1      | 2022-04-28T03:49:58.909672425Z [2022/04/28 03:49:58.909 +00:00] [INFO] [server.go:492] ["master client init successful"]
server-executor-0_1      | 2022-04-28T03:49:58.909690025Z [2022/04/28 03:49:58.909 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-0:10240]
server-executor-0_1      | 2022-04-28T03:49:58.914160528Z [2022/04/28 03:49:58.912 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-1:10240]
server-executor-0_1      | 2022-04-28T03:49:58.914175528Z [2022/04/28 03:49:58.913 +00:00] [INFO] [client.go:85] ["add new server master client"] [addr=server-master-2:10240]
server-executor-0_1      | 2022-04-28T03:49:58.926941837Z [2022/04/28 03:49:58.925 +00:00] [INFO] [server.go:516] ["resource client init successful"]
server-executor-0_1      | 2022-04-28T03:49:58.941497148Z [2022/04/28 03:49:58.938 +00:00] [INFO] [server.go:535] ["register successful"] [info="{\"type\":2,\"id\":\"\",\"addr\":\"server-executor-0:10241\",\"cap\":100}"]
server-executor-0_1      | 2022-04-28T03:49:58.941514848Z [2022/04/28 03:49:58.939 +00:00] [INFO] [server.go:385] ["listen address"] [addr=0.0.0.0:10241]
server-executor-0_1      | 2022-04-28T03:49:58.951489155Z [2022/04/28 03:49:58.948 +00:00] [INFO] [server.go:412] ["update service discovery metastore"] [addr=server-master-1:10240]
server-executor-0_1      | 2022-04-28T03:49:58.958789161Z [2022/04/28 03:49:58.958 +00:00] [INFO] [server.go:450] ["update framework metastore"] [addr=frame-etcd-standalone:2379]
server-executor-0_1      | 2022-04-28T03:49:58.961948763Z [2022/04/28 03:49:58.960 +00:00] [INFO] [server.go:[47](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:47)3] ["update user metastore"] [addr=user-etcd-standalone:2379]
server-executor-0_1      | 2022-04-28T03:49:59.012936401Z [2022/04/28 03:49:59.012 +00:00] [INFO] [discovery_keepalive.go:76] ["add peer"] [uuid=master0-cf07b10e-fa86-4723-9bc7-4053ace41ccf] [exec="{\"type\":1,\"id\":\"master0-cf07b10e-fa86-4723-9bc7-4053ace41ccf\",\"addr\":\"server-master-0:10240\",\"cap\":0}"]
server-executor-0_1      | 2022-04-28T03:49:59.012976501Z [2022/04/28 03:49:59.012 +00:00] [INFO] [discovery_keepalive.go:76] ["add peer"] [uuid=master1-339510d5-c198-4e8e-a258-83bfaa0cab72] [exec="{\"type\":1,\"id\":\"master1-339510d5-c198-4e8e-a258-83bfaa0cab72\",\"addr\":\"server-master-1:10240\",\"cap\":0}"]
server-executor-0_1      | 2022-04-28T03:49:59.013009601Z [2022/04/28 03:49:59.012 +00:00] [INFO] [discovery_keepalive.go:76] ["add peer"] [uuid=master2-afdcf3bc-8566-4d71-8a14-abc570a24d2c] [exec="{\"type\":1,\"id\":\"master2-afdcf3bc-8566-4d71-8a14-abc570a24d2c\",\"addr\":\"server-master-2:10240\",\"cap\":0}"]
server-executor-0_1      | 2022-04-28T03:49:59.013081601Z [2022/04/28 03:49:59.013 +00:00] [INFO] [discovery_keepalive.go:76] ["add peer"] [uuid=61f2dcee-96d4-4f9e-b3e1-cff311fc56f6] [exec="{\"type\":2,\"id\":\"61f2dcee-96d4-4f9e-b3e1-cff311fc56f6\",\"addr\":\"server-executor-1:10241\",\"cap\":100}"]
server-executor-0_1      | 2022-04-28T03:49:59.[48](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:48)5628547Z [2022/04/28 03:[49](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:49):59.480 +00:00] [WARN] [server.go:583] ["heartbeat response meet error"] [code=UnknownExecutor]
Error: rver-executor-0_1      | 2022-04-28T03:49:59.485660847Z [2022/04/28 03:49:59.481 +00:00] [ERROR] [http.go:28] ["debug server returned"] [error="mux: server closed"]
Error: rver-executor-0_1      | 2022-04-28T03:49:59.485666447Z [2022/04/28 03:49:59.480 +00:00] [ERROR] [server.go:137] ["message server exited"] []
Error: rver-executor-0_1      | 2022-04-28T03:49:59.485699347Z [2022/04/28 03:49:59.481 +00:00] [ERROR] [main.go:68] ["run executor with error"] [error="[DFLOW:ErrHeartbeat]logic error: [DFLOW:ErrUnknownExecutorID]cannot find executor ID: "] [errorVerbose="[DFLOW:ErrHeartbeat]logic error: [DFLOW:ErrUnknownExecutorID]cannot find executor ID: \ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:155\ngithub.com/hanfei1991/microcosm/executor.(*Server).keepHeartbeat\n\t/dataflow-engine/executor/server.go:586\ngithub.com/hanfei1991/microcosm/executor.(*Server).Run.func4\n\t/dataflow-engine/executor/server.go:359\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.0.0-202102200329[51](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:51)-036812b2e83c/errgroup/errgroup.go:[57](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:57)\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd[64](https://github.com/hanfei1991/microcosm/runs/6204649187?check_suite_focus=true#step:7:64).s:1571"]
```

--- a/executor/server.go
+++ b/executor/server.go
@@ -542,6 +542,7 @@ func (s *Server) selfRegister(ctx context.Context) (err error) {
 		retry.WithMaxTries(15 /* fail after 33 seconds, TODO: make it configurable */),
 		retry.WithIsRetryableErr(func(err error) bool {
 			if err.Error() == pb.ErrorCode_MasterNotReady.String() {
+				log.L().Info("server master leader is not ready, retry later")
 				return true
 			}
 			return false

--- a/executor/server.go
+++ b/executor/server.go
@@ -533,8 +533,6 @@ func (s *Server) selfRegister(ctx context.Context) (err error) {
 			return err2
 		}
 		if resp.Err != nil {
-			log.L().Info("register failed",
-				zap.String("error", resp.Err.Code.String()), zap.String("error2", pb.ErrorCode_MasterNotReady.String()))
 			return pcErrors.New(resp.Err.Code.String())
 		}
 		return nil

--- a/executor/server.go
+++ b/executor/server.go
@@ -5,8 +5,10 @@ import (
 	"strings"
 	"time"
 
+	pcErrors "github.com/pingcap/errors"
 	"github.com/pingcap/tiflow/dm/pkg/log"
 	p2pImpl "github.com/pingcap/tiflow/pkg/p2p"
+	"github.com/pingcap/tiflow/pkg/retry"
 	"github.com/pingcap/tiflow/pkg/security"
 	"github.com/pingcap/tiflow/pkg/tcpserver"
 	"go.etcd.io/etcd/client/pkg/v3/logutil"
@@ -522,10 +524,36 @@ func (s *Server) selfRegister(ctx context.Context) (err error) {
 		Address:    s.cfg.AdvertiseAddr,
 		Capability: defaultCapability,
 	}
-	resp, err := s.masterClient.RegisterExecutor(ctx, registerReq, s.cfg.RPCTimeout)
+
+	var resp *pb.RegisterExecutorResponse
+	err = retry.Do(ctx, func() error {
+		var err2 error
+		resp, err2 = s.masterClient.RegisterExecutor(ctx, registerReq, s.cfg.RPCTimeout)
+		if err2 != nil {
+			return err2
+		}
+		if resp.Err != nil {
+			log.L().Info("register failed",
+				zap.String("error", resp.Err.Code.String()), zap.String("error2", pb.ErrorCode_MasterNotReady.String()))
+			return pcErrors.New(resp.Err.Code.String())
+		}
+		return nil
+	},
+		retry.WithBackoffBaseDelay(200 /* 200 ms */),
+		retry.WithBackoffMaxDelay(3000 /* 3 seconds */),
+		retry.WithMaxTries(15 /* fail after 33 seconds, TODO: make it configurable */),
+		retry.WithIsRetryableErr(func(err error) bool {
+			if err.Error() == pb.ErrorCode_MasterNotReady.String() {
+				return true
+			}
+			return false
+		}),
+	)
+
 	if err != nil {
-		return err
+		return
 	}
+
 	s.info = &model.NodeInfo{
 		Type:       model.NodeTypeExecutor,
 		ID:         model.ExecutorID(resp.ExecutorId),

--- a/executor/server_test.go
+++ b/executor/server_test.go
@@ -10,12 +10,16 @@ import (
 	"time"
 
 	"github.com/phayes/freeport"
+	"github.com/pingcap/errors"
 	"github.com/pingcap/tiflow/dm/pkg/log"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
 	"google.golang.org/grpc"
 
+	"github.com/hanfei1991/microcosm/client"
 	"github.com/hanfei1991/microcosm/executor/worker"
+	"github.com/hanfei1991/microcosm/pb"
+	"github.com/hanfei1991/microcosm/pkg/uuid"
 )
 
 func init() {
@@ -121,4 +125,67 @@ func testCustomedPrometheusMetrics(t *testing.T, addr string) {
 		metric := string(body)
 		return strings.Contains(metric, "dataflow_executor_task_num")
 	}, time.Second, time.Millisecond*20)
+}
+
+type registerExecutorReturnValue struct {
+	resp *pb.RegisterExecutorResponse
+	err  error
+}
+
+type mockRegisterMasterClient struct {
+	client.MasterClient
+	respChan chan *registerExecutorReturnValue
+}
+
+func newMockRegisterMasterClient(chanBufferSize int) *mockRegisterMasterClient {
+	return &mockRegisterMasterClient{
+		respChan: make(chan *registerExecutorReturnValue, chanBufferSize),
+	}
+}
+
+func (c *mockRegisterMasterClient) RegisterExecutor(
+	ctx context.Context, req *pb.RegisterExecutorRequest, timeout time.Duration,
+) (resp *pb.RegisterExecutorResponse, err error) {
+	value := <-c.respChan
+	return value.resp, value.err
+}
+
+func TestSelfRegister(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	cfg := NewConfig()
+	port, err := freeport.GetFreePort()
+	require.Nil(t, err)
+	addr := fmt.Sprintf("127.0.0.1:%d", port)
+	cfg.AdvertiseAddr = addr
+	s := NewServer(cfg, nil)
+	mockMasterClient := newMockRegisterMasterClient(10)
+	s.masterClient = mockMasterClient
+
+	mockMasterClient.respChan <- &registerExecutorReturnValue{
+		nil, errors.New("service unavailable"),
+	}
+	err = s.selfRegister(ctx)
+	require.Error(t, err, "service unavailable")
+
+	executorID := uuid.NewGenerator().NewString()
+	returnValues := []*registerExecutorReturnValue{
+		{
+			&pb.RegisterExecutorResponse{
+				Err: &pb.Error{Code: pb.ErrorCode_MasterNotReady},
+			}, nil,
+		},
+		{
+			&pb.RegisterExecutorResponse{
+				ExecutorId: executorID,
+			}, nil,
+		},
+	}
+	for _, val := range returnValues {
+		mockMasterClient.respChan <- val
+	}
+	err = s.selfRegister(ctx)
+	require.NoError(t, err)
+	require.Equal(t, executorID, string(s.info.ID))
 }

